entity-framework/core/querying/client-eval.md

```diff
@@ -1,69 +1,62 @@
-
-<!-- [!code-csharp[Main](samples/core/Querying/ClientEval/Sample.cs?highlight=6)] -->
-``` csharp
-var blogs = context.Blogs
-    .OrderByDescending(blog => blog.Rating)
-    .Select(blog => new
-    {
-        Id = blog.BlogId,
-        Url = StandardizeUrl(blog.Url)
-    })
-    .ToList();
-```
+> [!TIP]
+> You can view this article's [sample](https://github.com/aspnet/EntityFramework.Docs/tree/master/samples/core/Querying) on GitHub.
 
-<!-- [!code-csharp[Main](samples/core/Querying/ClientEval/Sample.cs)] -->
-``` csharp
-public static string StandardizeUrl(string url)
-{
-    url = url.ToLower();
+In the following example, a helper method is used to standardize URLs for blogs, which are returned from a SQL Server database. Since the SQL Server provider has no insight into how this method is implemented, it isn't possible to translate it into SQL. All other aspects of the query are evaluated in the database, but passing the returned `URL` through this method is done on the client.
 
-    if (!url.StartsWith("http://"))
-    {
-        url = string.Concat("http://", url);
-    }
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientProjection)]
 
-    return url;
-}
-```
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientMethod)]
 
-## Client evaluation performance issues
+## Unsupported client evaluation
 
-While client evaluation can be very useful, in some instances it can result in poor performance. Consider the following query, where the helper method is now used in a filter. Because this can't be performed in the database, all the data is pulled into memory and then the filter is applied on the client. Depending on the amount of data, and how much of that data is filtered out, this could result in poor performance.
+While client evaluation is useful, it can result in poor performance sometimes. Consider the following query, in which the helper method is now used in a where filter. Because filter can't be applied in the database, all the data needs to be pulled into memory to apply filter on the client. Based on the amount of data on the server and the filter, client evaluation could result in poor performance. So Entity Framework Core blocks such client evaluation and throws a runtime exception.
 
-<!-- [!code-csharp[Main](samples/core/Querying/ClientEval/Sample.cs)] -->
-``` csharp
-var blogs = context.Blogs
-    .Where(blog => StandardizeUrl(blog.Url).Contains("dotnet"))
-    .ToList();
-```
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientWhere)]
+
+## Explicit client evaluation
+
+User may need to force into client evaluation explicitly in certain cases like following
+
+- Data is small so that doing evaluation on client doesn't cause huge performance penalty.
```

Comment:
You can disagree with it but our users don't. Our users are indeed using `AsEnumerable` to cause client eval and it is part of our exception message too.

Reply:
I'm talking about your logic. The fact that something doesn't cause a huge perf penalty is not, in itself, a justification to use it. You use it because you have no choice or because there's some other benefit.

Reply:
Again that is not true. It is simple cost vs benefit analysis. If I have table of 40 rows and a client filter function, I can also convert the client function to a database function and use UDF to map it. I can get best performance at the expense of spending ours to generate UDF when I can just do client eval because it is just cheaper and does not cause huge penalty.